frontend/lib/util/clang-integration.cpp

```diff
@@ -357,88 +363,335 @@ createClangPrecompiledHeader(Context* context, ID externBlockId) {
+          formals.emplace_back(formalName, UntypedFnSignature::DK_NO_DEFAULT,
+                               /* decl */ nullptr);
+
+          const clang::Type* formalClangType =
+              clangFormal->getType().getTypePtr();
+          auto formalChplType =
+              convertClangTypeToChapelType(context, formalClangType);
+          auto intent = formalClangType->isArrayType() ? QualifiedType::REF
+                                                       : QualifiedType::IN;
+          formalChplType = QualifiedType(intent, formalChplType.type());
+          formalTypes.push_back(formalChplType);
+        }
+
+        const UntypedFnSignature* untypedSig = UntypedFnSignature::get(
+            context, fnId, name,
+            /* isMethod */ false,
+            /* isTypeConstructor */ false,
+            /* isCompilerGenerated */ true,
+            /* throws */ false,
+            /* idTag */ uast::asttags::Function,
+            /* kind */ uast::Function::Kind::PROC, std::move(formals),
+            /* whereClause */ nullptr,
+            /* compilerGeneratedOrigin */ fnId);
+
+        result = TypedFnSignature::get(
+            context, untypedSig, std::move(formalTypes),
+            /* whereClauseResult */ TypedFnSignature::WHERE_NONE,
+            /* needsInstantiation */ false,
+            /* instantiatedFrom */ nullptr,
+            /* parentFn */ nullptr,
+            /* formalsInstantiated */ Bitmap(),
+            /* outerVariables */ OuterVariables());
+      }
+    }
+  });
+#endif
+
+  return QUERY_END(result);
+}
+
+static const QualifiedType& precompiledHeaderRetTypeForFnQuery(
+    Context* context, const TemporaryFileResult* pch, UniqueString name) {
+  QUERY_BEGIN(precompiledHeaderRetTypeForFnQuery, context, pch, name);
+
+  QualifiedType result;
+
+#ifdef HAVE_LLVM
+  runFuncOnIdent(context, pch, name, [&](const clang::Decl* decl) {
+    if (decl) {
+      if (auto fnDecl = llvm::dyn_cast<clang::FunctionDecl>(decl)) {
+        const clang::Type* clangReturnType =
+            fnDecl->getReturnType().getTypePtr();
+        auto chplReturnType =
+            convertClangTypeToChapelType(context, clangReturnType);
+        result = chplReturnType;
+      }
+    }
+  });
+#endif
+
+  return QUERY_END(result);
+}
+
 bool precompiledHeaderContainsName(Context* context,
                                    const TemporaryFileResult* pch,
                                    UniqueString name) {
   return precompiledHeaderContainsNameQuery(context, pch, name);
 }
 
+bool precompiledHeaderContainsFunction(Context* context,
+                                       const TemporaryFileResult* pch,
+                                       UniqueString name) {
+  return precompiledHeaderContainsFunctionQuery(context, pch, name);
+}
+
+const QualifiedType& precompiledHeaderTypeForSymbol(
+    Context* context, const TemporaryFileResult* pch, UniqueString name) {
+  return precompiledHeaderTypeForSymbolQuery(context, pch, name);
+}
```

Comment:
Usually we distinguish between the "query" and the "regular" function to hide the fact that it returns an internal const ref. If you're going to expose the const ref, I don't see the need for a wrapper.

Though, below, you do use a pointer to return TypedFnSignature, which DOES require a wrapper.